psdisk/psdisk.c

```diff
@@ -387,199 +338,188 @@ static int psdisk_updateImg(void)
 
 static int psdisk_parseToRm(void)
 {
+	struct plist_node_t *node;
 	int i;
-	struct plist_node_t *pNode;
 
 	/* Parse partition name */
-	for (i = 0; i < strlen(optarg); ++i) {
+	for (i = 0; i < strlen(optarg); i++) {
 		if (!isalnum(optarg[i])) {
-			fprintf(stderr,"Invalid partition name - %s.\n", optarg);
+			fprintf(stderr, "Invalid partition name - %s.\n", optarg);
 			return -1;
 		}
 	}
 
-	if ((pNode = (struct plist_node_t *)calloc(1, sizeof(struct plist_node_t))) == NULL) {
-		fprintf(stderr,"Cannot allocate memory, err: %s.\n", strerror(errno));
+	node = calloc(1, sizeof(struct plist_node_t));
+	if (node == NULL) {
+		fprintf(stderr, "Cannot allocate memory, err: %s.\n", strerror(errno));
 		return -1;
 	}
 
-	if (strlen(optarg) >= sizeof(pNode->pHeader.name)) {
-		fprintf(stderr,"Invalid partition name - %s.\n", optarg);
-		free(pNode);
+	if (strlen(optarg) >= sizeof(node->part.name)) {
+		fprintf(stderr, "Invalid partition name - %s.\n", optarg);
+		free(node);
 		return -1;
 	}
 	else {
-		strncpy((char *)pNode->pHeader.name, optarg, sizeof(pNode->pHeader.name));
+		strncpy((char *)node->part.name, optarg, sizeof(node->part.name));
 	}
 
-	pNode->status = part_remove;
-
-	LIST_INSERT_HEAD(&psdisk_common.list, pNode, ptrs);
-	++psdisk_common.pCnt;
+	node->status = part_remove;
+	LIST_INSERT_HEAD(&psdisk_common.list, node, ptrs);
+	psdisk_common.count++;
 
 	return 0;
 }
 
 
 static int psdisk_parseToSave(void)
 {
-	int i, ret = 0;
-	char *argsStr, *endptr;
+	struct plist_node_t *node = NULL;
 	const char *delimiter = ",";
-	struct plist_node_t *pNode = NULL;
-
-	const int MAX_PARTITION_OPTS = 4;
-	char *opts[MAX_PARTITION_OPTS];
+	char *args, *endptr;
+	char *opts[4];
+	int i, ret = 0;
 
 	/* Get data from optarg and separete them */
-	if ((argsStr = (char *)malloc(strlen(optarg) + 1)) == NULL) {
-		fprintf(stderr,"Cannot allocate memory, err: %s.\n", strerror(errno));
+	args = malloc(strlen(optarg) + 1);
+	if (args == NULL) {
+		fprintf(stderr, "Cannot allocate memory, err: %s.\n", strerror(errno));
 		return -1;
 	}
 
 	do {
-		strcpy(argsStr, optarg);
-		if ((opts[0] = strtok(argsStr, delimiter)) == NULL) {
-			fprintf(stderr,"Invalid number of arguments.\n");
+		strcpy(args, optarg);
+		if ((opts[0] = strtok(args, delimiter)) == NULL) {
+			fprintf(stderr, "Invalid number of arguments.\n");
 			break;
 		}
 
-		for (i = 1; i < MAX_PARTITION_OPTS; ++i) {
+		for (i = 1; i < sizeof(opts) / sizeof(opts[0]); i++) {
 			if ((opts[i] = strtok(NULL, delimiter)) == NULL) {
-				fprintf(stderr,"Invalid number of arguments.\n");
+				fprintf(stderr, "Invalid number of arguments.\n");
 				ret = 1;
 				break;
 			}
 		}
```

Comment:
Using `strtok()` in this case is not good solution, better `strsep()`.
And probably the best without spliting using condition `*endptr != ','` and only to the last `*endptr != '\0'`, immediately we have a check of number of parameters.
Such `optarg` will be correct `,,,p1,,2,,3,4,,,,5`.
You do not check number of parameters.
Suggestion:
```c

		for (i = 0; i < sizeof(opts) / sizeof(opts[0]); i++) {
			opts[i] = strtok(i == 0 ? args : NULL, delimiter);
			if (opts[i] == NULL) {
				fprintf(stderr, "Invalid number of arguments.\n");
				ret = 1;
				break;
			}
		}
```

Reply:
I agree that `strtok()` usage here seems unnecessary. I've changed the code to use suggested `*endptr != ','` and `*endptr != '\0'`. Applied in the next force-push.